test/e2e/firewall.go

```diff
@@ -0,0 +1,410 @@
+	"k8s.io/kubernetes/pkg/kubectl"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	compute "google.golang.org/api/compute/v1"
+)
+
+const (
+	firewallTimeoutDefault  = 3 * time.Minute
+	firewallTestHttpTimeout = time.Duration(1 * time.Second)
+	// Set ports outside of 30000-32767, 80 and 8080 to avoid being whitelisted by the e2e cluster
+	firewallTestHttpPort = int32(29999)
+	firewallTestUdpPort  = int32(29998)
+)
+
+var _ = framework.KubeDescribe("Firewall rule", func() {
+	var firewall_test_name = "firewall-test"
+	f := framework.NewDefaultFramework(firewall_test_name)
+
+	// This test takes around 3 minutes to run
+	It("[slow] [serial] should create valid firewall rules for LoadBalancer type service", func() {
+		framework.SkipUnlessProviderIs("gce")
+		c := f.ClientSet
+		ns := f.Namespace.Name
+		cloudConfig := framework.TestContext.CloudConfig
+		gceCloud := framework.TestContext.CloudConfig.Provider.(*gcecloud.GCECloud)
+		// This source ranges is just used to examine we have the same things on LB firewall rules
+		firewallTestSourceRanges := []string{"0.0.0.0/1", "128.0.0.0/1"}
+
+		// OnlyLocal service is needed to examine which exact nodes the requests are being forward to by the Load Balancer on GCE
+		By("Creating a LoadBalancer type service with onlyLocal annotation")
+		serviceName := "firewall-test-loadbalancer"
+		jig := NewServiceTestJig(c, serviceName)
+		svc := createOnlyLocalLoadBalancServiceWithPort(jig, ns, serviceName, firewallTestHttpPort, firewallTestSourceRanges)
+		defer func() {
+			Expect(c.Core().Services(svc.Namespace).Delete(svc.Name, nil)).NotTo(HaveOccurred())
+		}()
+		svcExternalIP := svc.Status.LoadBalancer.Ingress[0].IP
+
+		By("Checking if service's firewall rules are correct")
+		Expect(examineSvcFirewallRules(gceCloud, svc)).NotTo(HaveOccurred())
+
+		By("Creating a daemon set of netexec pods ")
+		dsName := "firewall-test-daemon-set"
+		Expect(createNetexecDaemonSet(c, dsName, ns, jig.Labels, firewallTestHttpPort, firewallTestUdpPort, true)).NotTo(HaveOccurred())
+		defer func() {
+			deleteDaemonSet(f, dsName, ns, jig.Labels)
+		}()
+
+		By("Checking that daemon pods launch on every node of the cluster.")
+		Expect(wait.Poll(dsRetryPeriod, dsRetryTimeout, checkRunningOnAllNodes(f, jig.Labels))).NotTo(HaveOccurred(),
+			"error waiting for daemon pod to start")
+		Expect(checkDaemonStatus(f, dsName)).NotTo(HaveOccurred())
+
+		// Send requests from outside of the cluster because internal traffic is whitelisted
+		By("Accessing the external service ip from outside, all non-master nodes should be reached")
+		nodeNames, err := getNonMaterNodeNames(c)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(examineHitNodesFromOutside(svcExternalIP, firewallTestHttpPort, firewallTimeoutDefault, nodeNames, true)).NotTo(HaveOccurred())
+
+		By("Removing tags from one of the nodes")
```

Comment:
please add a comment here about exactly what we're trying to verify. Something like
```
check if there are overlapping tags on the firewall that extend beyond just the vms in our cluster, 
by removing the tag on one vm and make sure it doesn't get any traffic. This is an imperfect 
simulation, we really want to check that traffic doesn't reach a vm outside the GKE cluster, but 
that's much harder to do in the current e2e framework. 
```